src/misc/lv_array.c

```diff
@@ -45,71 +45,143 @@ void lv_array_init(lv_array_t * array, uint32_t capacity, uint32_t element_size)
+    if(index == array->size - 1) {
+        lv_array_resize(array, array->size - 1);
+        return false;
+    }
+
+    uint8_t * start = lv_array_at(array, index);
+    uint8_t * remaining = start + array->element_size;
+    uint32_t remaining_size = (array->size - index - 1) * array->element_size;
+    lv_memcpy(start, remaining, remaining_size);
+    lv_array_resize(array, array->size - 1);
+    return true;
+}
+
+bool lv_array_erase(lv_array_t * array, uint32_t start, uint32_t end)
 {
-    if(new_capacity > array->size) {
-        if(new_capacity > array->capacity) {
-            uint8_t * data = lv_malloc(new_capacity * array->element_size);
-            lv_memcpy(data, array->data, array->size * array->element_size);
-            lv_free(array->data);
-            array->data = data;
-            array->capacity = new_capacity;
-        }
+    if(end > array->size) {
+        end = array->size;
+    }
+
+    if(start >= end) {
+        return false;
+    }
+
+    /*Shortcut*/
+    if(end == array->size) {
+        lv_array_resize(array, start);
+        return true;
     }
-    else {
+
+    uint8_t * start_p = lv_array_at(array, start);
+    uint8_t * remaining = start_p + (end - start) * array->element_size;
+    uint32_t remaining_size = (array->size - end) * array->element_size;
+    lv_memcpy(start_p, remaining, remaining_size);
+    lv_array_resize(array, array->size - (end - start));
+    return true;
+}
+
+void lv_array_resize(lv_array_t * array, uint32_t new_capacity)
+{
+    uint8_t * data = lv_realloc(array->data, new_capacity * array->element_size);
+    LV_ASSERT_NULL(data);
+    array->data = data;
+    array->capacity = new_capacity;
+    if(array->size > new_capacity) {
         array->size = new_capacity;
     }
 }
 
-bool lv_array_append(lv_array_t * array, const uint8_t * element)
+bool lv_array_concat(lv_array_t * array, const lv_array_t * other)
 {
-    if(array->size >= array->capacity) {
-        // array is full
+    LV_ASSERT_NULL(array->data);
+    uint32_t size = other->size;
+    if(array->size + size > array->capacity) {
+        /*array is full*/
         return false;
     }
 
+    uint8_t * data = array->data + array->size * array->element_size;
+    lv_memcpy(data, other->data, array->element_size * size);
+    array->size += size;
+    return true;
+}
+
+bool lv_array_push_back(lv_array_t * array, const void * element)
+{
     LV_ASSERT_NULL(array->data);
-    uint8_t * data = (uint8_t *)(array->data + array->size * array->element_size);
+
+    if(array->size == array->capacity) {
+        /*array is full*/
+        return false;
+    }
+
+    uint8_t * data = array->data + array->size * array->element_size;
     lv_memcpy(data, element, array->element_size);
     array->size++;
     return true;
 }
 
-uint8_t * lv_array_get(const lv_array_t * array, uint32_t index)
+void * lv_array_at(const lv_array_t * array, uint32_t index)
 {
-    if(index > (array->size - 1)) {
+    if(index >= array->capacity) {
```

Comment:
it should be: `array->size`
or do you want to allow devs to access unused elements in the array? (indexes between `size` and `capacity`)

using `array->capacity` would lead to a "bug":
if an array has 4 elements, and the capacity is 8, then the following call would allow to assign value at an unused index, and the size of array still remains 4, so assigned element would be still outside of used range:
`lv_array_assign(array, 5, "value");`

Reply:
I intend to do this way because I saw the original test case relies on this feature(with buggy code). I'm not sure if we should leave this backdoor open. I does can simplify some situations.

Note, 
```c
void test_array_set_get(void)
{
    uint32_t v = 100;
    lv_array_assign(&array, 0, &v);
    uint32_t * r = lv_array_at(&array, 0);
    TEST_ASSERT_EQUAL_UINT32(100, *r);
}
```

Original buggy code, note when size is zero, it won't return NULL.

```c
uint8_t * lv_array_get(const lv_array_t * array, uint32_t index)
{
    if(index > (array->size - 1)) {
        return NULL;
    }

    LV_ASSERT_NULL(array->data);
    return array->data + index * array->element_size;
}
```

Reply:
I have updated to use `array->size ` instead. Test cases updated too.
Returning pointer to non-existent element could be dangerous. For now, there's no such case need this feature.

Using `array->size` also matches with STL.